---
title: "Skin tape striping data analysis"
author: "Tu Hu"
date: "12/13/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
pacman::p_load("dplyr", "stringr")
```

## Mapping

```{r fastq files}
sample <-
  openxlsx::read.xlsx("../data/stp_sorting.xlsx") %>% 
  as_tibble()

fn <- 
  list.files("../data/fastq/", pattern = ".fastq.gz") %>% 
  str_remove("_R(1|2)_001.fastq.gz") %>% 
  unique()
writeLines(fn, "../_tmp/fn.txt")

batch <-
  fn %>% str_sub(1,1)
writeLines(batch, "../_tmp/batch.txt")  
```

We received in total `r length(fn)` * 2 (paired-end) files.

### Batch information
```{r}
batch_info <- 
  openxlsx::read.xlsx("../data/batch_info.xlsx") %>% 
  as_tibble() %>% 
  filter(PCR.batch != "-")
batch_info %>% 
  filter(PCR.batch == "PCR1") %>% 
  select(Name = sample_id, 
         B1 = UMI.primer.sequence) %>% 
  readr::write_tsv("../_tmp/batch_info_1.txt")
```


### QC

```{bash fastqc}
module load parallel/20200522 jdk/17 perl/5.30.2 fastqc/0.11.9
cd ..
ID_PATH=_tmp/fn.txt
cat ${ID_PATH} | parallel "fastqc data/fastq/{}_R1_001.fastq.gz data/fastq/{}_R2_001.fastq.gz -o data/fastqc"

```

```{bash multiqc}
module load anaconda3/4.4.0
multiqc ../data/fastqc -o ../data/fastqc
```

```{r average sequencing depth}
total_read <- 
  readr::read_tsv("../data/fastqc/multiqc_data/multiqc_fastqc.txt") %>% 
  select(Sample, `Total Sequences`) %>% 
  mutate(Sample = Sample %>% str_remove("_R(1|2)_001")) %>% 
  distinct() %>% 
  summarise(sum_read = sum(`Total Sequences`)) %>% 
  pull(sum_read)

total_read / nrow(sample)
```

### Construct reference genome

First, download the latest reference genome from Ensembl

```{bash download ref genome from ensemble}
mkdir -p ../data/reference
wget -P ../data/reference ftp://ftp.ensembl.org/pub/release-105/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
gzip -d ../data/reference/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
wget -P ../data/reference ftp://ftp.ensembl.org/pub/release-105/gtf/homo_sapiens/Homo_sapiens.GRCh38.105.gtf.gz
gzip -d ../data/reference/Homo_sapiens.GRCh38.105.gtf.gz
```

```{bash}
```


```{bash}
module load jre/1.8.0 brbseqtools/1.6.0
```

```{bash constract ref genome}
module load star/2.7.9a
mkdir -p ../data/STAR_Index/
cd ../data
STAR --runMode genomeGenerate --genomeDir STAR_Index/ --genomeFastaFiles reference/Homo_sapiens.GRCh38.dna.primary_assembly.fa --sjdbGTFfile reference/Homo_sapiens.GRCh38.105.gtf --runThreadN 30

```

```{bash STAR alignment reverse read}
cd ../data
module load star/2.7.9a parallel/20210722
fn_path=../_tmp/fn.txt
cat ${fn_path} | parallel "STAR --runMode alignReads --genomeDir STAR_Index/ --outFilterMultimapNmax 1 --readFilesCommand zcat --outSAMtype BAM Unsorted --outFileNamePrefix BAM/{} --readFilesIn fastq/{}_R2_001.fastq.gz --runThreadN 8"
```

```{bash}
cd ../data
module load jre/1.8.0 brbseqtools/1.6.0
brbseqtools CreateDGEMatrix -f fastq/1_S1_L001_R1_001.fastq.gz -b BAM/testAligned.out.bam -c ../_tmp/batch_info_1.txt -gtf reference/Homo_sapiens.GRCh38.105.gtf -p B???????????????????? -o ../_tmp/test
```

```{r}
test_mapping <- 
  readr::read_tsv("../_tmp/test/output.dge.reads.detailed.txt")
test_mapping[,-1:-2] %>% colSums()
```

```{r}
test_mapping %>% 
  select(Gene_id, Gene_name, 
         `05_AD_12_LS`, 
         `05_AD_22_LS`, `02_AD_22_NL`) %>% 
  arrange(- `05_AD_12_LS`) %>% 
  write.csv("../_tmp/mapping_3_samples.csv")
```

