---
title: "Skin tape striping data analysis"
author: "Tu Hu"
date: "12/13/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
pacman::p_load("dplyr", "stringr", "purrr", "tidyr", "ggplot2")
```

## Mapping

```{r fastq files}
sample <-
  openxlsx::read.xlsx("../data/stp_sorting.xlsx") %>% 
  as_tibble()

fn <- 
  list.files("../data/fastq/", pattern = ".fastq.gz") %>% 
  str_remove("_R(1|2)_001.fastq.gz") %>% 
  unique()
writeLines(fn, "../_tmp/fn.txt")

batch <-
  fn %>% str_sub(1,1)
writeLines(batch, "../_tmp/batch.txt")  
```

We received in total `r length(fn)` * 2 (paired-end) files.

### Batch information
```{r}
batch_info <- 
  openxlsx::read.xlsx("../data/batch_info.xlsx") %>% 
  as_tibble() %>% 
  filter(PCR.batch != "-") %>% 
  group_by(PCR.batch) %>% 
  select(Name = sample_id, 
         B1 = UMI.primer.sequence) %>% 
  tidyr::nest() %>% 
  mutate(fn = paste0("../_tmp/batch_info_", PCR.batch %>% str_extract("\\d{1}"), ".txt"))

purrr::map2(batch_info$data,
            batch_info$fn, ~ readr::write_tsv(.x, .y))
```


### QC

```{bash fastqc}
module load parallel/20200522 jdk/17 perl/5.30.2 fastqc/0.11.9
cd ..
ID_PATH=_tmp/fn.txt
cat ${ID_PATH} | parallel "fastqc data/fastq/{}_R1_001.fastq.gz data/fastq/{}_R2_001.fastq.gz -o data/fastqc"

```

```{bash multiqc}
module load anaconda3/4.4.0
multiqc ../data/fastqc -o ../data/fastqc
```

```{r average sequencing depth}
total_read <- 
  readr::read_tsv("../data/fastqc/multiqc_data/multiqc_fastqc.txt") %>% 
  select(Sample, `Total Sequences`) %>% 
  mutate(Sample = Sample %>% str_remove("_R(1|2)_001")) %>% 
  distinct() %>% 
  summarise(sum_read = sum(`Total Sequences`)) %>% 
  pull(sum_read)

total_read / nrow(sample)
```

### Construct reference genome

First, download the latest reference genome from Ensembl

```{bash download ref genome from ensemble}
mkdir -p ../data/reference
wget -P ../data/reference ftp://ftp.ensembl.org/pub/release-105/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
gzip -d ../data/reference/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
wget -P ../data/reference ftp://ftp.ensembl.org/pub/release-105/gtf/homo_sapiens/Homo_sapiens.GRCh38.105.gtf.gz
gzip -d ../data/reference/Homo_sapiens.GRCh38.105.gtf.gz
```

```{bash}
```


```{bash}
module load jre/1.8.0 brbseqtools/1.6.0
```

```{bash constract ref genome}
module load star/2.7.9a
mkdir -p ../data/STAR_Index/
cd ../data
STAR --runMode genomeGenerate --genomeDir STAR_Index/ --genomeFastaFiles reference/Homo_sapiens.GRCh38.dna.primary_assembly.fa --sjdbGTFfile reference/Homo_sapiens.GRCh38.105.gtf --runThreadN 30
```

```{bash STAR alignment reverse read}
module load gcc/9.4.0 star/2.7.9a parallel/20210722
cd /home/projects/ku_00015/people/tuhu/multiomics-ad-sts/data
fn_path=../_tmp/fn.txt
for i in $(cat ${fn_path}); do STAR --genomeDir STAR_Index/ \
--runMode alignReads \
--outFilterMultimapNmax 1 \
--readFilesCommand zcat \
--outSAMtype BAM Unsorted \
--outFileNamePrefix BAM/$i. \
--readFilesIn fastq/$i\_R2_001.fastq.gz \
--runThreadN 20 ; done
```

```{bash use forward read to map samples}
cd ../data
module load jre/1.8.0 brbseqtools/1.6.0 parallel/20210722
create_brbmatrix(){
brbseqtools CreateDGEMatrix -f fastq/$1_R1_001.fastq.gz -b BAM/$1.Aligned.out.bam -c ../_tmp/batch_info_$2.txt -gtf reference/Homo_sapiens.GRCh38.105.gtf -p B???????????????????? -o $1
}
export -f create_brbmatrix
parallel --link -a ../_tmp/fn.txt -a ../_tmp/batch.txt create_brbmatrix
```

##
```{r}
reads <- 
  tibble(fn = list.dirs("../data/") %>% grep("\\d{1}_S\\d{1}_L00\\d{1}$", ., value = T)) %>% 
  mutate(read = paste0(fn, "/output.dge.reads.detailed.txt"),
         read = map(read, readr::read_tsv, show_col_types=FALSE),
         lib_size = map(read, ~ .x[,-1:-2] %>% colSums),
         batch = fn %>% str_extract("\\d{1}(?=\\_S)"),
         lane = fn %>% str_extract("L00\\d{1}"),
         sample_n = map(lib_size, names))

lib_size <- reads %>% select(lib_size, batch) %>% group_by(batch) %>% nest() %>% 
  mutate(lib_size = map(data, ~ .x[[1]] %>% reduce(`+`)))

libsize_g <- 
  reads %>% select(batch, lane, lib_size, sample_n) %>% 
  unnest() %>% 
  ggplot(aes(sample_n)) + geom_bar(aes(weight=lib_size)) +
  facet_grid(lane~batch, scales = "free") + scale_y_log10() + 
  theme(axis.text.x = element_text(angle = 90))
ggsave("../_tmp/counts_sample.jpg", libsize_g, width = 21, height = 7)

counts <-
  reads %>% select(batch, read) %>% 
  mutate(read = map(read, ~ .x[,-1:-2])) %>% 
  group_by(batch) %>% nest() %>% 
  mutate(read = map(data, ~ .x[[1]] %>% reduce(`+`))) %>% 
  pull(read) %>% reduce(cbind) %>% 
  cbind(reads$read[[1]][,1:2], .) %>% as_tibble(.name_repair = "unique")
         # read = map(read, ~ cbind(reads$read[[1]][,1:2], .x)))

```

```{r}
g <- counts$Gene_name
genes <- c(counts$Gene_name)

require(biomaRt)
mart <- useMart("ENSEMBL_MART_ENSEMBL")
mart <- useDataset("hsapiens_gene_ensembl", mart)
annotLookup <- getBM(
  mart = mart,
  attributes = c(
    "hgnc_symbol",
    "entrezgene_id",
    "ensembl_gene_id",
    "gene_biotype"),
  filter = "hgnc_symbol",
  values = g,
  uniqueRows=TRUE)
g_p <- annotLookup %>% filter(gene_biotype == "protein_coding") %>% pull(hgnc_symbol)

counts_g_p <- counts %>% filter(Gene_name %in% g_p)
counts_g_p %>% readr::write_csv("../data/counts_pcoding.csv")
```

