---
title: "Skin tape striping - data QC and curation"
author: "Tu Hu"
date: "1/7/2022"
output: html_document
---

## Mapping

Bulk RNA barcoding and sequencing (BRB-seq) data was mapped by aligning reverse reads to human reference genome (GRCh38.105) by **STAR(2.7.9a)**, then de-multiplexing by the forward read barcoding. 
The average mapping rate is **`r knitr::load_cache("load-reads", "mean_map_rate")`**, which is comparable with conventional RNA-seq (such as GSE193309).

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
pacman::p_load(
  "dplyr", "stringr", "purrr", "tidyr", "ggplot2", 
  "tibble", "SummarizedExperiment", "tidySummarizedExperiment", 
  "tidybulk", "ComplexHeatmap", "ggpubr", "DT",
  "DESeq2", "BiocParallel", "edgeR", "lme4")
pca_color <- c("LS" = "#eb2d0c", "NL" = "#eb8b9b", "HC" = "#91cf60")
core_n <- future::availableCores()
register(MulticoreParam(ifelse(core_n <= 8, core_n - 2, core_n / 2)))
```

```{r load-reads, message=FALSE, warning=FALSE, cache=TRUE}
# star results
calc_map_rate <- 
  function(df = ...){
    sum(df[nrow(df)-4:nrow(df),-1:-2]) / sum(df[,-1:-2])
  }

reads <- 
  tibble(path = list.dirs("../data/brbmatrix") %>% grep("\\d{1}$", ., value = T)) %>% 
  mutate(read = paste0(path, "/output.dge.reads.detailed.txt"),
         read = map(read, readr::read_tsv, show_col_types=FALSE),
         lib_size = map(read, ~ .x[,-1:-2] %>% colSums),
         sample_n = map(lib_size, names),
         batch = path %>% str_extract("\\d{1}$"),
         map_rate = map_dbl(read, calc_map_rate))

mean_map_rate <-
  scales::label_percent()(mean(reads$map_rate))
```

```{r combine reads from five batches}
counts <-
  reads %>% 
  dplyr::select(batch, read) %>% 
  mutate(read = map(read, ~ .x[,-1:-2])) %>% 
  group_by(batch) %>% nest() %>% 
  mutate(read = map(data, ~ .x[[1]] %>% purrr::reduce(`+`))) %>% 
  pull(read) %>% purrr::reduce(cbind) %>% 
  cbind(reads$read[[1]][,1:2], .) # gene name and gene id

counts_filterS <- 
  counts[,!counts %>% names() %>%
           str_detect("\\.|Unknown_Barcode|02_CO_33")] %>% 
  as_tibble() %>% 
  filter(!is.na(Gene_name))

```

```{r, eval=FALSE}
# MT genes
MT_ratio <- 
  tibble(
  MT_counts = counts[counts$Gene_name %>% str_detect("MT\\-") %>% which, -1:-2] %>% colSums(),
  total_counts = counts[, -1:-2] %>% colSums(),
  MT_ratio = (MT_counts / total_counts) * 100
) %>% 
  arrange(MT_ratio) %>% select(MT_ratio) %>% 
  mutate(rank = 1:nrow(.)) %>% 
  ggplot(aes(rank, MT_ratio)) + geom_point() +
  ylab("counts(MT genes) / total counts") +
  theme_classic() +
  ggtitle("Ratio of mitochondrial gene counts and total counts")

MT_c <-
  tibble(
  MT_counts = counts[counts$Gene_name %>% str_detect("MT\\-") %>% which, -1:-2] %>% colSums(),
  total_counts = counts[, -1:-2] %>% colSums(),
  MT_ratio = (MT_counts / total_counts) * 100
) %>% 
  ggplot(aes(total_counts, MT_counts)) + geom_point() +
  theme_classic() +
  ggtitle("Ratio of mitochondrial gene counts and total counts")
  
  
ggsave("../data/figure/MT_ratio.jpeg", MT_ratio, height = 5)

```

```{r keep only protein coding genes, message=FALSE, warning=FALSE, eval=FALSE}
g <- counts$Gene_name
genes <- c(counts$Gene_name)
mart <- biomaRt::useMart("ENSEMBL_MART_ENSEMBL")
mart <- biomaRt::useDataset("hsapiens_gene_ensembl", mart)
annotLookup <- biomaRt::getBM(
  mart = mart,
  attributes = c(
    "hgnc_symbol",
    "entrezgene_id",
    "ensembl_gene_id",
    "gene_biotype"),
  filter = "hgnc_symbol",
  values = g,
  uniqueRows=TRUE)
g_p <- annotLookup %>% filter(gene_biotype == "protein_coding") %>% pull(hgnc_symbol)
saveRDS(g_p, file = "../data/g_p.rds") # time-consuming to exec
```

```{r filter only p coding g}
g_p <- readRDS("../data/g_p.rds") # time-consuming to exec
# g_p <- g_p[!g_p %>% str_detect("^MT-")] # remove MT genes
counts_g_p <- counts_filterS %>% filter(Gene_name %in% g_p)
```

## Metadata

(Skipped ... same as GSE193309)

```{r clinical record, message=FALSE, warning=FALSE}
se_punch_biopsy_rnaseq_raw <-
  readr::read_rds("../../multiomics-ad-phd/data/se_raw.rds")

meta_punch_biopsy_rnaseq_raw <- 
  colData(se_punch_biopsy_rnaseq_raw) %>% 
  as_tibble() %>% 
  mutate(sample_name_join = paste(visit, subject, skin_type, sep = "_")) %>% 
  dplyr::select(sample_name_join, 
         gender, contains("scorad"), easi_total_score,
         date_visit, visit_quarter, biopsy_area) %>% 
  distinct() %>% filter(!is.na(visit_quarter), 
                        !is.na(biopsy_area),
                        !(sample_name_join == "05_AD_13_LS" & date_visit == "21.01.2019"),
                        !(sample_name_join == "03_AD_13_NL" & date_visit == "01.11.2019"))

metadata <- 
  tibble(sample_name = colnames(counts_g_p)[-1:-2],
         visit = sample_name %>% str_extract("^\\d{2}"),
         group = sample_name %>% str_extract("AD|CO"),
         subject_id = sample_name %>% str_extract("(AD|CO)_\\d{2}"),
         skin_type = sample_name %>% str_extract("LS|NL")) %>% 
  mutate(skin_type = ifelse(group == "CO", "HC", skin_type),
         sample_name = paste(visit, subject_id, skin_type, sep = "_")) %>%  
  left_join(meta_punch_biopsy_rnaseq_raw, by=c("sample_name" = "sample_name_join")) %>% 
  mutate(date_visit = as.Date(date_visit, "%d.%m.%Y"))

colnames(counts_g_p)[-1:-2] <- metadata$sample_name
# counts_g_p %>% readr::write_csv("../data/counts_pcoding.csv")

# openxlsx::write.xlsx(metadata, "../data/metadata.xlsx")

metadata_edit <- 
  readr::read_csv("../data/metadata_edit.csv") %>% 
  mutate(date_visit = as.Date(date_visit, "%m/%d/%Y"))

metadata_new <-
  metadata %>% dplyr::select(sample_name) %>% left_join(metadata_edit) %>% 
  mutate(skin_type = skin_type %>% forcats::fct_relevel(c("HC", "NL", "LS")),
         visit_quarter = visit_quarter %>% forcats::fct_relevel(c("Q1", "Q2", "Q3", "Q4")))
```

```{r output data for Thomas Litman, eval=FALSE}
metadata_new %>% readr::write_csv("../_tmp/tape_striping_metadata.csv")
counts_g_p %>% readr::write_csv("../_tmp/tape_striping_counts.csv")
```

```{r construct se}
se <- SummarizedExperiment(
  assays = list(counts = counts_g_p[,-1:-2] %>% as.matrix()),
  colData = metadata_new
)
names(se) <- counts_g_p$Gene_name

counts_sum <- 
  se %>% assay() %>% colSums() %>% enframe(name = "sample", value = "counts_sum")

```

## QC

We applied a 4-point data QC procedures. The samples score 1-point when fulfilling each criteria: (1) the sensitivity of detection (detecting more than 200 genes with higher than 1 count). (2) epidermal differentiation complex (EDC) gene signal intensities, as measured by FLG, FLG2, LORICRIN, LCE1A, S100A7, S100A8, SPRR2E, SPRR1B, IVL. (3) EDC genes is comprised > 5% of total counts (4) mitochondrial genes is comprised <15% of total counts.

### Wrong labels

```{r filtering out wrongly labelled sample}
wrong_labels <-
  counts %>% names() %>% 
  grep("\\.|02_CO_33", ., value = T) %>% str_remove("\\.\\.\\.\\d{1,}") %>% table
```

We have some wrongly labelled samples, including:

```{r showing wrong labels}
wrong_labels
```

### Low read depth

*Read depth (of a sample) = sum of reads of all mapped genes*

```{r Visualize library size, message=FALSE, warning=FALSE}
lib_size <- 
  reads %>% dplyr::select(lib_size, batch) %>% unnest(lib_size)

lib_size_g <- 
  lib_size %>% 
  gghistogram(x = "lib_size", xscale = "log10", 
              binwidth = 0.3, xlab = "Library size (Read depth)", 
              title = "Histogram of library size (read depth) \nn(sample) = 207, n(gene)=17000", 
              add = "median", rug = TRUE, fill = "lightgray") 

lib_size_g
# ggsave("../data/figure/lib_size_g.jpeg", lib_size_g, width = 7, height = 5)

lib_size_g %>% facet("batch") %>%  ggpar(font.xtickslab = "8")

median_lib_size <- median(lib_size$lib_size)
```

We detected a low library sized in general, with median library size `r median_lib_size` and a large variation between RNA yield from each sample, as indicated by the total counts of each sample. However, between each BRB batch, we could not detect a batch effect. 

#### RNA yield (estimated by total counts) variation
When evaluating the variation of RNA yield, we observed an increasing trend of RNA yield in LS > NL > HC. LS has significantly higher RNA yield from HC (p < 0.001). 
We also observed the variations of RNA yield across quarters (Q1-Q4).
We tested RNA yield variation using linear mixed effect model.

```{r RNA yield - skin type}
colData(se)$counts_sum <- counts_sum$counts_sum

rna_yield_skin_type_g <- 
  se %>% pivot_sample() %>% 
  ggboxplot(x = "skin_type", 
            y = "counts_sum", 
            yscale = "log10",
            add = c("jitter"), 
            color = "skin_type",
            palette = pca_color,
            xlab = "Skin type",
            ylab = "Total counts") +
  stat_compare_means(comparisons = list(
    c("HC", "NL"),
    c("NL", "LS"),
    c("HC", "LS")
  ), methods= "t.test") +
  theme(legend.position = "none")
# ggsave("../figure/RNA_yield_skin_type.jpeg", 
#        width = 5, height = 5, units = "in", 
#        rna_yield_skin_type_g)
```

```{r RNA yield - season}
rna_yield_season_g <- 
  se %>% pivot_sample() %>% 
  filter(!is.na(visit_quarter)) %>% 
  ggboxplot(x = "visit_quarter", 
            y = "counts_sum", 
            yscale = "log10",
            add = c("jitter"), 
            color = "visit_quarter",
            xlab = "Quarter",
            ylab = "Total counts") +
    stat_compare_means(comparisons = list(
    c("Q1", "NL"),
    c("NL", "LS"),
    c("HC", "LS")
    ), methods= "t.test")+
  stat_compare_means(method = "anova") +
  theme(legend.position = "none")
# ggsave("../figure/RNA_yield_season.jpeg", 
#        width = 5, height = 5, units = "in", 
#        rna_yield_season_g)
```

```{r rna-yield-lme}
rna_yield_lme <-
  lmerTest::lmer(
  counts_sum ~ skin_type + visit_quarter + biopsy_area + (1|subject_id), 
  data = se %>% pivot_sample() ) %>% broom.mixed::tidy()
```

### Unknown barcode

```{r unknown barcode, cache=TRUE}
unknown_barcode_percent <- 
  sum(counts[-(nrow(counts)-4) : -(nrow(counts)), colnames(counts) == "Unknown_Barcode"]) / sum(counts[-(nrow(counts)-4) : -(nrow(counts)), -1:-2])
```

Within mapped reads, `r scales::label_percent()(unknown_barcode_percent)` of reads can't find their origins (detected as unknown barcode).

## Data curation

```{r filter out low expr genes}
se_x <- # se expressed
  se[rowMeans(assay(se) > 0) > 0.2,]
```

```{r total counts and sensitivity}
qc_total_counts <- se_x %>% assay() %>% colSums()
qc_sensitivity_g_0 <- colSums(assay(se_x) > 0) # no. of detected genes int. > 0 
qc_sensitivity_g_1 <- colSums(assay(se_x) > 1) # no. of detected genes int. > 1 
qc_sensitivity_g_2 <- colSums(assay(se_x) > 2) # no. of detected genes int. > 2
```

```{r MT genes}
g_MT <- rownames(se_x) %>% grep("MT-", ., value = T)
qc_MT_ratio <- (assay(se_x)[g_MT,] %>% colSums())*100/qc_total_counts
qc_MT_ratio15 <- qc_MT_ratio < 15
```

```{r hair genes}
g_hair <- c("KRTAP1-5",
            "KRTAP3-2",
            "KRTAP3-3",
            "KRTAP4-1",
            "KRTAP11-1")
qc_hair_1percent <- (colSums(assay(se_x[g_hair,])) / qc_total_counts) > .01 
```

```{r EDC genes}
g_EDC <- c("FLG", "FLG2",
           "LORICRIN", "LCE1A", 
           "S100A8","S100A7",
           "SPRR2E","SPRR1B",
           "IVL")
qc_EDC_5percent <- (colSums(assay(se_x[g_EDC,])) / qc_total_counts) > .05 # percentage of EDC within all counts
qc_EDC_count100 <- colSums(assay(se_x[g_EDC,])) > 100
```

```{r construct new se}
se_qc <- se_x[!rownames(se_x) %>% str_detect("MT-"), ] #2310
se_qc$qc_hair_1percent <- qc_hair_1percent
se_qc$qc_MT_ratio15 <- qc_MT_ratio15
se_qc$qc_EDC_5percent <- qc_EDC_5percent
se_qc$qc_EDC_count100 <- qc_EDC_count100
se_qc$qc_sensitivity_200 <- qc_sensitivity_g_1 > 200
se_qc$qc_0_4 <- qc_EDC_5percent + qc_EDC_count100 + (qc_sensitivity_g_1 > 200) + qc_MT_ratio15
se_qc$qc_coverage <- ifelse(se_qc$counts_sum < 35000, "low", "high")
# saveRDS(se_qc, "../data/se_qc.rds")


se_qc_wMT <- se_x[,] #2323
colData(se_qc_wMT) <- colData(se_qc)
# saveRDS(se_qc_wMT, "../data/se_qc_wMT.rds")
```

After QC, we detected expression of `r nrow(se_qc[,se_qc$qc_0_4>2])` genes, with > 1 count in at least 80% of samples. After a 4-point scheme quality filtering, We kept `r ncol(se_qc[,se_qc$qc_0_4>2])` samples for further analysis.

### Statistics on filtered samples

```{r eval=FALSE}
# se_qc <- readRDS("../data/se_qc.rds")
```

```{r}
se_qc %>% pivot_sample() %>% 
  group_by(skin_type) %>% 
  summarise(n = n())
```

```{r qc statistics}
se_qc %>% pivot_sample() %>% 
  filter(qc_0_4 < 3) %>% 
  group_by(skin_type) %>% 
  summarise(n = n())
```

```{r lmer}
se_qc_vp <- se_qc[,se_qc$qc_0_4 > 2]
colData(se_qc_vp)$RNA_yield <- 
  se_qc_vp %>% 
  assay() %>% 
  colSums() %>% 
  log2()
se_qc_vp_col <- se_qc_vp %>% pivot_sample()

lmerTest::lmer(RNA_yield ~ gender + skin_type + visit_quarter + qc_hair_1percent + (1|subject_id), se_qc_vp_col) %>% broom.mixed::tidy()
```

```{r}
se_qc_vp %>% pivot_transcript() %>% 
  mutate(median_expr = se_qc_vp %>% assay() %>% matrixStats::rowMeans2()) %>% 
  slice_max(median_expr, n = 10)
```

```{r}
se_qc_vp[, se_qc_vp$skin_type %in% c("NL", "HC")] %>% 
  scale_abundance(method = "TMMwsp") %>% 
  pivot_transcript() %>% 
  mutate(median_expr = se_qc_vp %>% assay() %>% rowMeans()) %>% 
  slice_max(median_expr, n = 10)
```
```{r}
gene_detection <- 
  se_qc_vp[, se_qc_vp$skin_type %in% c("NL", "HC")] %>% 
  scale_abundance(method = "TMMwsp") %>% 
  pivot_transcript() %>% 
  mutate(median_expr = se_qc_vp %>% assay() %>% rowMeans()) %>% 
  arrange(-median_expr)

which(gene_detection$feature == "KRT10")
which(gene_detection$feature == "KRT14")
which(gene_detection$feature == "KRTAP1-5")
which(gene_detection$feature == "KRT5")
which(gene_detection$feature == "KRTAP4-9")
which(gene_detection$feature == "FLG2")
which(gene_detection$feature == "KRT86")
which(gene_detection$feature == "KRT17")
which(gene_detection$feature == "ACTG1")
```

```{r enrichr}
gene_detection$feature[1:500] %>% writeLines("../_tmp/gene_expressed.txt")
```


## Heatmap

## GLMPCA
```{r glmpca}
### PCA plot using generalized PCA
library("glmpca")
se_qc <- readr::read_rds("../data/se_qc.rds")
gpca <- glmpca(assay(se_qc), L = 2, fam = "nb2")
gpca.dat <- gpca$factors


se_qc[table$`Gene (overexpressed)` %>% unlist %>% unique, ] %>% 
  pivot_sample() %>% 
  left_join(gpca$factors %>% as_tibble(rownames = "sample")) %>% 
  ggplot(aes(x = dim1, y = dim2, color = skin_type)) +
  geom_point() +
  theme_classic()
  # scale_color_manual()


se_qc_dr <- se_qc %>% reduce_dimensions(method = "PCA")

pca_plot <-
  se_qc_dr %>% 
  pivot_sample() %>% 
  rename(`Sequencing coverage` = qc_coverage,
         `Tissue type` = skin_type) %>% 
  ggscatter(x = "PC1", y = "PC2", color = "Sequencing coverage", 
            shape = "Tissue type", font.legend = c(10, "plain", "black"))

ggsave("../data/pca.jpeg", pca_plot, width = 6, height = 5)
```



